EventGenerator/src/CaloCalibGun_module.cc

```diff
@@ -0,0 +1,284 @@
+      _Ntupe -> Branch("genErg", &_genErg, "genErg/F");
+      _Ntupe -> Branch("genTime", &_genTime, "genTime/F");
+      _Ntupe -> Branch("genCos", &_genCos, "genCos/F");
+      _Ntupe -> Branch("genPosX", &_genPosX, "genPosX/F");
+      _Ntupe -> Branch("genPosY", &_genPosY, "genPosY/F");
+      _Ntupe -> Branch("genPosZ", &_genPosZ, "genPosZ/F");
+    }
+    if (_mean < 0) throw cet::exception("RANGE") << "CaloCalibGun.mean must be non-negative "<< '\n';
+  }
+
+  //================================================================
+  void CaloCalibGun::produce(art::Event& event) {
+    std::unique_ptr<GenParticleCollection> output(new GenParticleCollection);
+    PrimaryParticle primaryParticles;
+    MCTrajectoryCollection mctc;
+
+    _cal =  &*GeomHandle<mu2e::DiskCalorimeter>();
+    //_nPipes          = _cal->caloInfo().getInt("nPipes");
+    _pipeRadius      = _cal->caloInfo().getDouble("pipeRadius");
+    _pipeTorRadius   = _cal->caloInfo().getVDouble("pipeTorRadius");
+    _randomRad       = _cal->caloInfo().getVDouble("pipeTorRadius");
+    _zPipeCenter     = _cal->disk(1).geomInfo().origin()-CLHEP::Hep3Vector(0,0,_cal->disk(1).geomInfo().size().z()/2.0-_pipeRadius);//disk =1
+    // we normalize to the volume of the pipe (proportional to 2*pi*R if they have all the same radius) to draw a random number from which to generate the photons TODO - is this used?
+    double sumR(0);
+    std::for_each(_randomRad.begin(), _randomRad.end(), [&](double& d) {sumR+=d; d = sumR; });
+    std::for_each(_randomRad.begin(), _randomRad.end(), [&](double& d) {d /= sumR;});
+    
+    unsigned int nGen = _mean;
+    //Define the parameters of the pipes:
+    
+    // angle of large torus in degrees
+    double phi_lbd[5] = {161.34, 149.50, 139.50, 132.07, 125.39};
+    // angle of small torus in degrees
+    double phi_sbd[5] = {84.63, 85.28, 85.79, 86.20, 86.53};
+    // angle of the end point
+    double phi_end[5] = {3.96, 10.53, 15.80, 20.16, 23.84};
+    // center position y of the small torus
+    double ysmall[5] = {432.2, 480.5, 524.3, 564.7, 602.5};
+    // radius of small torus
+    double radSmTor = 41.0;
+    // first center position x of the small torus
+    double xsmall = 71.0;
+    // distance of the small torus center
+    double xdistance = 60.0;
+    // inner radius of the manifold
+    double rInnerManifold = 681.6; // 713.35 mm - 1.25 in (31.75 mm)
+    vector<float> sign{-1.0, 1.0};
+
+    for (unsigned int ig = 0; ig < nGen; ++ig) 
+    {
+      // only in the front of 2nd disk.
+      //int nDisk        = 1;
+      //_zPipeCenter     = _cal->disk(nDisk).geomInfo().origin()-CLHEP::Hep3Vector(0,0,_cal->disk(nDisk).geomInfo().size().z()/2.0-_pipeRadius);
+
+      double xpipe, ypipe, zpipe;
+      //Pick position
+      int xsn = rint(_randFlat.fire());
+      int ysn = rint(_randFlat.fire());
+
+      double theta = _randFlat.fire() * 2.0 * piconst;
+      double pipeR = _pipeRadius * _randFlat.fire();
```

Comment:
I am not sure what you are trying to do here.  A distribution that is uniform across a unit disk is not uniform in r.  Or are you doing something else?

Reply:
It should be uniform in R as the pipes are at set radii

Reply:
If you want to sample a uniform position inside a disk to generate the starting position of a photon. then Rob is right, a uniform distribution in a disk is given by 
R * sqrt(_randFlat.fire())
otherwise you over-sample at the center. If you want to sample something else, then I'm not sure what.